File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] arr) {
        int n = arr.length;
        int[] right = new int[n];

        int prod = 1;

        for(int i = n - 1; i >= 0; i--) {
            prod = prod * arr[i];
            right[i] = prod;
        } 

        int[] ans = new int[n];

        int left = 1;

        for(int i = 0; i < n-1; i++) {
            int val = left * right[i+1];

            ans[i] = val;

            left = left * arr[i];

        }

        ans[n-1] = left;

        return ans;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> pre(n);
        vector<int> suf(n);
        vector<int> ans(n);

        // prefix product array
        int p = nums[0];
        pre[0] = 1;
        for(int i = 1; i < n; i++) {
            pre[i] = p;
            p *= nums[i];
        }

        // suffix product array
        p = nums[n-1];
        suf[n-1] = 1;
        for(int i = n-2; i >=0 ; i--) {
            suf[i] = p;
            p *= nums[i];
        }

        // ans product array Except self
        for(int i = 0; i < n; i++) {
            ans[i] = pre[i] * suf[i];
        }

        return ans;
    }
};